File: sdk/src/main/java/com/cxense/cxensesdk/DeviceInfoProvider.java
package com.cxense.cxensesdk;

import android.content.Context;
import android.content.pm.PackageManager;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.telephony.TelephonyManager;
import android.util.DisplayMetrics;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2018-09-17).
 */
class DeviceInfoProvider {
    private static final String TAG = DeviceInfoProvider.class.getSimpleName();
    private final Context context;

    DeviceInfoProvider(@NonNull Context context) {
        this.context = context;
    }

    DisplayMetrics getDisplayMetrics() {
        return context.getResources().getDisplayMetrics();
    }

    /**
     * Get version of the application in which SDK is used.
     *
     * @return application's version or 'null' if not found
     */
    @Nullable
    String getApplicationVersion() {
        try {
            return context.getPackageManager().getPackageInfo(context.getPackageName(), 0).versionName;
        } catch (PackageManager.NameNotFoundException e) {
            Log.e(TAG, "Problems during application version search", e);
        }
        return null;
    }

    /**
     * Get name of the application in which SDK is used.
     *
     * @return application's name or 'null' if not found
     */
    @Nullable
    String getApplicationName() {
        CharSequence label = context.getPackageManager().getApplicationLabel(context.getApplicationInfo());
        return label != null ? label.toString() : null;
    }

    /**
     * Gets current network status
     *
     * @return {@code NetworkStatus} instance
     */
    CxenseConfiguration.NetworkStatus getCurrentNetworkStatus() {
        ConnectivityManager manager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        final NetworkInfo activeNetworkInfo = manager.getActiveNetworkInfo();

        if (activeNetworkInfo == null || !activeNetworkInfo.isConnected()) {
            return CxenseConfiguration.NetworkStatus.NONE;
        }
        if (activeNetworkInfo.getType() == ConnectivityManager.TYPE_WIFI) {
            return CxenseConfiguration.NetworkStatus.WIFI;
        }
        TelephonyManager telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
        switch (telephonyManager.getNetworkType()) {
            case TelephonyManager.NETWORK_TYPE_HSDPA:
            case TelephonyManager.NETWORK_TYPE_HSPA:
            case TelephonyManager.NETWORK_TYPE_HSPAP:
            case TelephonyManager.NETWORK_TYPE_HSUPA:
            case TelephonyManager.NETWORK_TYPE_LTE:
                return CxenseConfiguration.NetworkStatus.MOBILE;
            default:
                return CxenseConfiguration.NetworkStatus.GPRS;
        }
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/DeviceInfoProvider.kt
package com.cxense.cxensesdk

import android.content.Context
import android.content.pm.PackageManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.telephony.TelephonyManager
import android.util.DisplayMetrics
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import timber.log.Timber

/**
 * Provides device info
 */
class DeviceInfoProvider(
    private val context: Context
) {
    val displayMetrics: DisplayMetrics by lazy { context.resources.displayMetrics }
    /**
     * Get version of the application in which SDK is used.
     *
     * @return application's version or 'null' if not found
     */
    val applicationVersion: String? by lazy {
        try {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName
        } catch (e: PackageManager.NameNotFoundException) {
            Timber.e(e, "Problems during application version search")
            null
        }
    }
    /**
     * Get name of the application in which SDK is used.
     *
     * @return application's name
     */
    val applicationName: String by lazy {
        context.packageManager.getApplicationLabel(context.applicationInfo).toString()
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun NetworkCapabilities.toNetworkStatus(): CxenseConfiguration.NetworkStatus {
        return when {
            hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                    hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> CxenseConfiguration.NetworkStatus.WIFI
            hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> CxenseConfiguration.NetworkStatus.MOBILE
            hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) -> CxenseConfiguration.NetworkStatus.GPRS
            else -> CxenseConfiguration.NetworkStatus.NONE
        }
    }

    /**
     * Gets current network status
     *
     * @return {@code NetworkStatus} instance
     */
    fun getCurrentNetworkStatus(): CxenseConfiguration.NetworkStatus {
        val manager = ContextCompat.getSystemService(context, ConnectivityManager::class.java)
            ?: return CxenseConfiguration.NetworkStatus.NONE

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            return manager.getNetworkCapabilities(manager.activeNetwork)?.toNetworkStatus()
                ?: CxenseConfiguration.NetworkStatus.NONE
        } else {
            @Suppress("DEPRECATION")
            val activeNetworkInfo = manager.activeNetworkInfo
            @Suppress("DEPRECATION")
            return when {
                activeNetworkInfo == null || !activeNetworkInfo.isConnected -> CxenseConfiguration.NetworkStatus.NONE
                activeNetworkInfo.type == ConnectivityManager.TYPE_WIFI -> CxenseConfiguration.NetworkStatus.WIFI
                else -> {
                    val telephonyManager =
                        ContextCompat.getSystemService(context, TelephonyManager::class.java)
                    when {
                        telephonyManager == null -> CxenseConfiguration.NetworkStatus.NONE
                        telephonyManager.networkType in MOBILE_NETWORK_TYPES -> CxenseConfiguration.NetworkStatus.MOBILE
                        else -> CxenseConfiguration.NetworkStatus.GPRS
                    }
                }
            }
        }
    }

    companion object {
        @JvmStatic
        val MOBILE_NETWORK_TYPES = arrayOf(
            TelephonyManager.NETWORK_TYPE_HSDPA,
            TelephonyManager.NETWORK_TYPE_HSPA,
            TelephonyManager.NETWORK_TYPE_HSPAP,
            TelephonyManager.NETWORK_TYPE_HSUPA,
            TelephonyManager.NETWORK_TYPE_LTE
        )
    }
}
